### PR TITLE
Add taskKillGracePeriod

### PR DESCRIPTION
**Pull request by kensipe**
Added ability to send taskKillGracePeriodSeconds to marathon libs from metronome.
Changed the name to include suffix `Seconds` which is consistent with parts of marathon:)

https://jira.mesosphere.com/browse/METRONOME-151

Jira:  METRONOME-151

### CODE REVIEW

**Comment by kensipe** [comment]
Looking into what it takes... manually tested.. needs unit tests

File: jobs/src/main/protobuf/metronome.proto [hunk 1]
```diff
@@ -173,6 +173,14 @@ message JobSpec {
...
     }
     optional RestartSpec restart = 13;
+
+    // mesos kill policy https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L625
+    // for some reason referred to as taskKillGracePeriod in marathon.
+    // https://github.com/mesosphere/marathon/blob/releases/1.3/src/main/scala/mesosphere/marathon/state/AppDefinition.scala#L79
+    // AND taskKillGracePeriodSeconds
+    // https://github.com/mesosphere/marathon/blob/releases/1.3/docs/docs/rest-api/public/api/v2/schema/AppDefinition.json
+    // int64 in mesos and int32 in marathon... going with int64
+    optional int64 taskKillGracePeriodSeconds = 14;
```
Comment (meichstedt) [inline jobs/src/main/protobuf/metronome.proto]: I'd stick with int32 – Mesos' `DurationInfo` specifies the amount in nanoseconds, that's why it's an int64.

File: jobs/src/main/scala/dcos/metronome/repository/impl/kv/marshaller/JobSpecMarshaller.scala [hunk 2]
```diff
@@ -136,6 +136,7 @@ object RunSpecConversions {
       runSpec.args.foreach { args => builder.addAllArguments(args.asJava) }
       runSpec.user.foreach(builder.setUser)
       runSpec.docker.foreach { docker => builder.setDocker(docker.toProto) }
+      runSpec.taskKillGracePeriodSeconds.foreach { killGracePeriod => builder.setTaskKillGracePeriodSeconds(killGracePeriod.toMillis) }
```
Comment (meichstedt) [inline jobs/src/main/scala/dcos/metronome/repository/impl/kv/marshaller/JobSpecMarshaller.scala]: Wait a _second_ – shouldn't this be **`toSeconds`**? 🙄 

Also, you're missing the conversion in `implicit class ProtoToRunSpec` on line 145ff.

(In the current impl the gracePeriod would be received via the API, stored in zk and passed over to Marathon, but not retrieved from zk after a failover).

File: jobs/src/main/protobuf/metronome.proto (hunk 1 above)
Comment (alenkacz) [reply jobs/src/main/protobuf/metronome.proto]: also I think that the code style for proto files require to use _ not camelCase. We should stick with that...

Comment (kensipe) [reply jobs/src/main/protobuf/metronome.proto]: all of our protobuf defs are camel... however I should have started with an upper case.

File: jobs/src/main/protobuf/metronome.proto [hunk 3]
```diff
@@ -173,6 +173,14 @@ message JobSpec {
...
     }
     optional RestartSpec restart = 13;
+
+    // mesos kill policy https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L625
+    // for some reason referred to as taskKillGracePeriod in marathon.
+    // https://github.com/mesosphere/marathon/blob/releases/1.3/src/main/scala/mesosphere/marathon/state/AppDefinition.scala#L79
+    // AND taskKillGracePeriodSeconds
+    // https://github.com/mesosphere/marathon/blob/releases/1.3/docs/docs/rest-api/public/api/v2/schema/AppDefinition.json
+    // int64 in mesos and int32 in marathon... going with int64
+    optional int64 TaskKillGracePeriodSeconds = 14;
```
Comment (alenkacz) [inline jobs/src/main/protobuf/metronome.proto]: oh well, now it is a weird mix, there is max_launch_delay but also activeDeadline ... I would personally stick with the official guidelines which is task_kill_grace...

Comment (kensipe) [reply jobs/src/main/protobuf/metronome.proto]: aw.. you're right

**Comment by alenkacz** [comment]
I think we should test:
- it is marshalled to proto (that one I already commited)
- it is properly parsed in controller
- it is passed to marathon (this one I am working on)

I think we don't need integration/SI test ... we can verify it is passed to marathon and we can expect that this is already tested in integration test in Marathon ... ? :)

I have a "JobRunExecutorActor" test already halfway ready - I will commit that one tomorrow.

Otherwise nice work @kensipe !

**Comment by alenkacz** [comment]
@kensipe I have submitted the tests so feel free to review if it makes sense to you like this

File: jobs/src/main/protobuf/metronome.proto [hunk 4]
```diff
@@ -173,6 +173,14 @@ message JobSpec {
...
     }
     optional RestartSpec restart = 13;
+
+    // mesos kill policy https://github.com/apache/mesos/blob/master/include/mesos/mesos.proto#L625
+    // for some reason referred to as taskKillGracePeriod in marathon.
+    // https://github.com/mesosphere/marathon/blob/releases/1.3/src/main/scala/mesosphere/marathon/state/AppDefinition.scala#L79
+    // AND taskKillGracePeriodSeconds
+    // https://github.com/mesosphere/marathon/blob/releases/1.3/docs/docs/rest-api/public/api/v2/schema/AppDefinition.json
+    // int64 in mesos and int32 in marathon... going with int64
+    optional int64 task_kill_grace_period_seconds = 14;
```
Comment (alenkacz) [inline jobs/src/main/protobuf/metronome.proto]: Matthias proposed to stick with int32 here. Not sure whether his comment was just lost or whether it was a decision to keep int64 so just pointing it out...

Comment (kensipe) [reply jobs/src/main/protobuf/metronome.proto]: int32 adds unnecessary complicity... we had a slack conversation which resulted in being ok to use int64

**Pull request merged**